zcash_primitives/src/transaction/builder.rs

```diff
@@ -691,14 +778,33 @@ impl<P: consensus::Parameters, R: RngCore + CryptoRng> Builder<P, R> {
         }
 
         // Add a binding signature if needed
-        if binding_sig_needed {
-            self.mtx.binding_sig = Some(
+        self.mtx.binding_sig = if binding_sig_needed {
+            Some(
                 prover
                     .binding_sig(&mut ctx, self.mtx.value_balance, &sighash)
-                    .map_err(|()| Error::BindingSig)?,
-            );
+                    .map_err(|_| Error::BindingSig)?,
+            )
         } else {
-            self.mtx.binding_sig = None;
+            None
+        };
+
+        // Create TZE input witnesses
+        for (i, tze_in) in self.tze_inputs.builders.into_iter().enumerate() {
+            // Need to enable witness to commit to the amount.
+            // - So hardware wallets "know" the amount without having to be sent all the
+            //   prior TZE outputs to which this witness gives evidence.
+            //
+            // The witness is expected to commit to the precommitment internally?
+            // (Or make it part of the sighash?)
+            // - TODO: Check whether transparent inputs committing to script_pubkey was
+            //   only so that hardware wallets "knew" what address was being spent from.
```

Comment:
@str4d what are your thoughts here? It seems like the information that a hardware wallet would need to display to the user in order for that user to be able to confidently sign a TZE transaction would vary by TZE; is there actually anything "general" that we can do here?

Reply:
I think the "general" thing we do is what we are currently doing: require the entire TZE precondition to be part of the signed transaction digest. Any future address format linked to TZEs would be inherently linked to a particular precondition, and thus derivable from it.

The "Need to enable witness to commit to the amount" is already "solved", because each TZE has access to the `prevout` before calling `Builder::add_tze_input`, so they can cache the needed properties there. Alternatively, we could explicitly pass the prevout back to them when calling `tze_in.builder`.

Reply:
I think expecting the TZE to cache what it needs is the proper thing to do here. That makes it so that there's a single path by which the implementers can achieve their goal, which is a good thing.